r/ecs - Add support for FARGATE Windows Containers and ARM based Containers

ChrisMcKee opened the pull request:
Adds platform runtime properties and validation based on https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html

```
  runtime_platform {
    operating_system_family = "WINDOWS_SERVER_2019_CORE"
    cpu_architecture = "X86_64"
  }
```

Adds documentation for new fields (but no new examples; I'd happily add examples for windows fargate if its wanted, I didn't want to bloat the document)


<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

Closes #21706
Closes #22033

Output from acceptance testing:

```
make testacc TESTS=TestAccECSTaskDefinition_withRuntimePlatform PKG=ecs
TF_ACC=1 go test ./internal/service/ecs/... -v -count 1 -parallel 20 -run='TestAccECSTaskDefinition_withRuntimePlatform' -timeout 180m
=== RUN   TestAccECSTaskDefinition_withRuntimePlatform
=== PAUSE TestAccECSTaskDefinition_withRuntimePlatform
=== CONT  TestAccECSTaskDefinition_withRuntimePlatform
--- PASS: TestAccECSTaskDefinition_withRuntimePlatform (14.22s)
PASS
ok  


make testacc TESTS=TestAccECSTaskDefinition_Fargate_withRuntimePlatform PKG=ecs
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecs/... -v -count 1 -parallel 20 -run='TestAccECSTaskDefinition_Fargate_withRuntimePlatform' -timeout 180m
=== RUN   TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== PAUSE TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== CONT  TestAccECSTaskDefinition_Fargate_withRuntimePlatform
--- PASS: TestAccECSTaskDefinition_Fargate_withRuntimePlatform (14.57s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ecs        14.620s
...
```

DrFaust92 commented on internal/service/ecs/task_definition.go:
```diff
@@ -143,6 +143,22 @@ func ResourceTaskDefinition() *schema.Resource {
 				ValidateFunc: validation.StringInSlice(ecs.NetworkMode_Values(), false),
 			},
 
+			"platform_family": {
```
I think this needs to follow the api and be a runtime_platform block containing these instead of having it flattened

ChrisMcKee replied:
I placed it based on platform family which was root based. Then the docs were half updated with this seemingly pointless extra object and platform family duplication 🤮

ChrisMcKee replied:
K I've done it based on the https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html
so its (lazily typed here)

runtime_platform {
  operating_system_family = "LINUX"
  cpu_architecture = "ARM64"
}

Just need to run the tests etc and I'll push that in

ChrisMcKee commented:
```shell
make testacc TESTS=TestAccECSTaskDefinition_withRuntimePlatform PKG=ecs
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecs/... -v -count 1 -parallel 20 -run='TestAccECSTaskDefinition_withRuntimePlatform' -timeout 180m
=== RUN   TestAccECSTaskDefinition_withRuntimePlatform
=== PAUSE TestAccECSTaskDefinition_withRuntimePlatform
=== CONT  TestAccECSTaskDefinition_withRuntimePlatform
--- PASS: TestAccECSTaskDefinition_withRuntimePlatform (24.44s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ecs        24.483s


make testacc TESTS=TestAccECSTaskDefinition_Fargate_withRuntimePlatform PKG=ecs
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecs/... -v -count 1 -parallel 20 -run='TestAccECSTaskDefinition_Fargate_withRuntimePlatform' -timeout 180m
=== RUN   TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== PAUSE TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== CONT  TestAccECSTaskDefinition_Fargate_withRuntimePlatform
--- PASS: TestAccECSTaskDefinition_Fargate_withRuntimePlatform (14.03s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ecs        14.075s

```

The time on the first ones higher as I ran the thing in US West (forgot to set my region env)

ChrisMcKee commented:
Installed the provider locally after the change and updated my test wireup 

Definition outputs fine
![task definition diff](https://user-images.githubusercontent.com/83597/144505125-9a23429a-c0d0-44dc-9d9d-877c077c95b1.png)

validations correct
![validation test](https://user-images.githubusercontent.com/83597/144505318-ec4aded0-1bf8-4f72-b437-fc6ce994a776.png)

![single runtime_platform restriction](https://user-images.githubusercontent.com/83597/144505469-db056345-62b4-49e2-95b3-0c1c2fbecc6e.png)

DrFaust92 commented on internal/service/ecs/task_definition.go:
```diff
@@ -143,6 +143,29 @@ func ResourceTaskDefinition() *schema.Resource {
+					Schema: map[string]*schema.Schema{
+						"operating_system_family": {
+							Type:         schema.TypeString,
+							Required:     true,
```
According to https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RuntimePlatform.html non are required

ChrisMcKee replied:
The other doc for the definitions said it was conditionally required if fargate was enabled 🤦‍♂️

ChrisMcKee replied:
ooi what's the difference between `Required: false` and `Optional: true`?

DrFaust92 replied:
There isnt AFAIK. the convention is not to use `Required: false` but rather be explicit with `Optional: true`

ChrisMcKee replied:
```
	// If one of these is set, then this item can come from the configuration.
	// Both cannot be set. If Optional is set, the value is optional. If
	// Required is set, the value is required.
	//
	// One of these must be set if the value is not computed. That is:
	// value either comes from the config, is computed, or is both.
	Optional bool
	Required bool
```

Threw me off; don't see many things like that in .net

DrFaust92 commented on internal/service/ecs/task_definition.go:
```diff
@@ -768,6 +814,18 @@ func expandEcsTaskDefinitionPlacementConstraints(constraints []interface{}) ([]*
+	config := runtimePlatformConfig[0]
+	configMap := config.(map[string]interface{})
+
+	ecsProxyConfig := &ecs.RuntimePlatform{
```
taking above into account only one of these (or even none) can provided. so lets only set each of the arguments if there are provided. this guard against nil defernce

ChrisMcKee replied:
Looking at `expandEcsTaskDefinitionRuntimePlatformConfiguration` the typelists is passed in as an interface and then as its an array of interface type we take the first entry... fine. But if you set neither property then the resulting entry is nil; but the cast to map[string]interface happens there after which would be an exception. 

I've added a nill check to return empty runtime_platform if the config variable is nil. But I'm wondering if I should as the rest of the similar objects dont. If you create a runtime_platform object with no properties in it, terraform plan shows the change each time

```
          # aws_ecs_task_definition.test must be replaced
              ~ revision                 = 1 -> (known after apply)
              ~ tags_all                 = {} -> (known after apply)
                # (5 unchanged attributes hidden)

              + runtime_platform { # forces replacement}
            }

        Plan: 1 to add, 0 to change, 1 to destroy.
```

I may be overthinking this as technically its not a valid config; but the config validation across ECS is already a bit vague.

ChrisMcKee replied:
Ignore that, thinking aloud.

DrFaust92 commented on internal/service/ecs/task_definition.go:
```diff
@@ -613,6 +645,20 @@ func flattenPlacementConstraints(pcs []*ecs.TaskDefinitionPlacementConstraint) [
+	}
+
+	config := make(map[string]interface{})
+	config["operating_system_family"] = aws.StringValue(pc.OperatingSystemFamily)
```
same as expand

ChrisMcKee replied:
I've done expand; which ironically worked empty anyway (going off my local test harness); the read-side though seems fine with emptys as things like ipc_mode are optional and theres no checks around these string values / StringValue protects against null 

```
func StringValue(v *string) string {
	if v != nil {
		return *v
	}
	return ""
}
```

DrFaust92 commented on internal/service/ecs/task_definition_test.go:
```diff
@@ -167,6 +167,62 @@ func TestAccECSTaskDefinition_withDockerVolumeMinimal(t *testing.T) {
+			{
+				Config: testAccTaskDefinitionWithRuntimePlatformMinimalConfig(tdName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckTaskDefinitionExists(resourceName, &def),
```
lets add checks for the values  for example: `resource.TestCheckResourceAttr(resourceName, "runtime_platform.#", "1"),`

DrFaust92 commented on internal/service/ecs/task_definition_test.go:
```diff
@@ -167,6 +167,62 @@ func TestAccECSTaskDefinition_withDockerVolumeMinimal(t *testing.T) {
+			{
+				Config: testAccFargateTaskDefinitionWithRuntimePlatformMinimalConfig(tdName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckTaskDefinitionExists(resourceName, &def),
```
same as above

ChrisMcKee replied:
K; added them in but I'm not sure I get the point in unit testing a string constant hcl record for a string?
Is it supposed to be testing the hcls parsed properly? Seems an odd thing for an integration test vs unit 🤷

DrFaust92 replied:
well, for starters this is not really a unit test. this tests the above code and integration with AWS E2E. but this is to check expected result equals the actual one.

ChrisMcKee replied:
Yeah tbh I was looking at the method names rather than digging into what each one was doing; it "looked" like it was just checking the string was in the hcl (i.e. checking that what was typed was typed which seemed pointless) 
But I see its checking the hcl as a map essentially checking the HCL will render (i pressume thats the intent)

The unit-test comment was solely about that section as its not hitting the API its just checking the terraform config format. Slows life down when you have to wait 14 seconds and you're blindly using someone elses code base filled with hundreds of helper methods you've never seen before 😄

ChrisMcKee commented:
```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecs/... -v -count 1 -parallel 20 -run='TestAccECSTaskDefinition_Fargate_withRuntimePlatform' -timeout 180m
=== RUN   TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== PAUSE TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== RUN   TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch
=== PAUSE TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch
=== CONT  TestAccECSTaskDefinition_Fargate_withRuntimePlatform           
=== CONT  TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch
--- PASS: TestAccECSTaskDefinition_Fargate_withRuntimePlatform (14.24s)
--- PASS: TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch (14.25s)     
PASS                                                                                   
ok      github.com/hashicorp/terraform-provider-aws/internal/service/ecs        14.298s
```

ChrisMcKee commented:
@DrFaust92 I tried the following out of interest; but that fails  (basically an empty `runtime_platform{}`) 
Terraform plan has content. The checks though in the read/write side are covering more than others like proxy config (they don't check empty).

Probably doesnt matter as its not technically valid config, just interested in why

```
func TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutProperties(t *testing.T) {
	var def ecs.TaskDefinition

	tdName := sdkacctest.RandomWithPrefix("tf-acc-td-with-runtime-platform")
	resourceName := "aws_ecs_task_definition.test"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:     func() { acctest.PreCheck(t) },
		ErrorCheck:   acctest.ErrorCheck(t, ecs.EndpointsID),
		Providers:    acctest.Providers,
		CheckDestroy: testAccCheckTaskDefinitionDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccFargateTaskDefinitionWithRuntimePlatformMinimalConfig(tdName, false, false),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckTaskDefinitionExists(resourceName, &def),
					resource.TestCheckResourceAttr(resourceName, "runtime_platform.#", "0"),
				),
			},
			{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateIdFunc: testAccTaskDefinitionImportStateIdFunc(resourceName),
				ImportStateVerify: true,
			},
		},
	})
}
```

maybejosh commented on internal/service/ecs/task_definition.go:
```diff
@@ -589,6 +617,10 @@ func resourceTaskDefinitionRead(d *schema.ResourceData, meta interface{}) error
 	}
 
+	if err := d.Set("runtime_platform", flattenRuntimePlatform(taskDefinition.RuntimePlatform)); err != nil {
+		return fmt.Errorf("error setting proxy_configuration: %w", err)
```
```suggestion
		return fmt.Errorf("error setting runtime_platform: %w", err)
```

ChrisMcKee replied:
Good catch thanks

ChrisMcKee commented:
Closes #22032

ChrisMcKee commented:
Cleaned up the pr and dealt with the hcl fmtcheck complaints (this repo really needs an editorconfig)

ChrisMcKee commented:
I squashed / cleaned up the pr and corrected some the action results that don't seem to be repeated when a prs updated in the checks here.  So thats all green now 
![image](https://user-images.githubusercontent.com/83597/144858377-9dcbc333-553c-4706-b5f6-5ed5b47537cf.png)

Pushed a separate PR with an editorconfig 😆  save my sanity next time maybe

ewbankkit reviewed (approved):
LGTM :rocket:.

```console
% make testacc TESTS=TestAccECSTaskDefinition_ PKG=ecs                                                
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/ecs/... -v -count 1 -parallel 20 -run='TestAccECSTaskDefinition_' -timeout 180m
=== RUN   TestAccECSTaskDefinition_basic
=== PAUSE TestAccECSTaskDefinition_basic
=== RUN   TestAccECSTaskDefinition_withScratchVolume
=== PAUSE TestAccECSTaskDefinition_withScratchVolume
=== RUN   TestAccECSTaskDefinition_withDockerVolume
=== PAUSE TestAccECSTaskDefinition_withDockerVolume
=== RUN   TestAccECSTaskDefinition_withDockerVolumeMinimal
=== PAUSE TestAccECSTaskDefinition_withDockerVolumeMinimal
=== RUN   TestAccECSTaskDefinition_withRuntimePlatform
=== PAUSE TestAccECSTaskDefinition_withRuntimePlatform
=== RUN   TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== PAUSE TestAccECSTaskDefinition_Fargate_withRuntimePlatform
=== RUN   TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch
=== PAUSE TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch
=== RUN   TestAccECSTaskDefinition_withEFSVolumeMinimal
=== PAUSE TestAccECSTaskDefinition_withEFSVolumeMinimal
=== RUN   TestAccECSTaskDefinition_withEFSVolume
=== PAUSE TestAccECSTaskDefinition_withEFSVolume
=== RUN   TestAccECSTaskDefinition_withTransitEncryptionEFSVolume
=== PAUSE TestAccECSTaskDefinition_withTransitEncryptionEFSVolume
=== RUN   TestAccECSTaskDefinition_withEFSAccessPoint
=== PAUSE TestAccECSTaskDefinition_withEFSAccessPoint
=== RUN   TestAccECSTaskDefinition_withFSxWinFileSystem
=== PAUSE TestAccECSTaskDefinition_withFSxWinFileSystem
=== RUN   TestAccECSTaskDefinition_withTaskScopedDockerVolume
=== PAUSE TestAccECSTaskDefinition_withTaskScopedDockerVolume
=== RUN   TestAccECSTaskDefinition_withECSService
=== PAUSE TestAccECSTaskDefinition_withECSService
=== RUN   TestAccECSTaskDefinition_withTaskRoleARN
=== PAUSE TestAccECSTaskDefinition_withTaskRoleARN
=== RUN   TestAccECSTaskDefinition_withNetworkMode
=== PAUSE TestAccECSTaskDefinition_withNetworkMode
=== RUN   TestAccECSTaskDefinition_withIPCMode
=== PAUSE TestAccECSTaskDefinition_withIPCMode
=== RUN   TestAccECSTaskDefinition_withPidMode
=== PAUSE TestAccECSTaskDefinition_withPidMode
=== RUN   TestAccECSTaskDefinition_constraint
=== PAUSE TestAccECSTaskDefinition_constraint
=== RUN   TestAccECSTaskDefinition_changeVolumesForcesNewResource
=== PAUSE TestAccECSTaskDefinition_changeVolumesForcesNewResource
=== RUN   TestAccECSTaskDefinition_arrays
=== PAUSE TestAccECSTaskDefinition_arrays
=== RUN   TestAccECSTaskDefinition_fargate
=== PAUSE TestAccECSTaskDefinition_fargate
=== RUN   TestAccECSTaskDefinition_Fargate_ephemeralStorage
=== PAUSE TestAccECSTaskDefinition_Fargate_ephemeralStorage
=== RUN   TestAccECSTaskDefinition_executionRole
=== PAUSE TestAccECSTaskDefinition_executionRole
=== RUN   TestAccECSTaskDefinition_disappears
=== PAUSE TestAccECSTaskDefinition_disappears
=== RUN   TestAccECSTaskDefinition_tags
=== PAUSE TestAccECSTaskDefinition_tags
=== RUN   TestAccECSTaskDefinition_proxy
=== PAUSE TestAccECSTaskDefinition_proxy
=== RUN   TestAccECSTaskDefinition_inferenceAccelerator
=== PAUSE TestAccECSTaskDefinition_inferenceAccelerator
=== CONT  TestAccECSTaskDefinition_basic
=== CONT  TestAccECSTaskDefinition_withNetworkMode
=== CONT  TestAccECSTaskDefinition_executionRole
=== CONT  TestAccECSTaskDefinition_withECSService
=== CONT  TestAccECSTaskDefinition_inferenceAccelerator
=== CONT  TestAccECSTaskDefinition_proxy
=== CONT  TestAccECSTaskDefinition_tags
=== CONT  TestAccECSTaskDefinition_withEFSVolume
=== CONT  TestAccECSTaskDefinition_disappears
=== CONT  TestAccECSTaskDefinition_withTaskRoleARN
=== CONT  TestAccECSTaskDefinition_withTaskScopedDockerVolume
=== CONT  TestAccECSTaskDefinition_withFSxWinFileSystem
=== CONT  TestAccECSTaskDefinition_withEFSAccessPoint
=== CONT  TestAccECSTaskDefinition_Fargate_ephemeralStorage
=== CONT  TestAccECSTaskDefinition_fargate
=== CONT  TestAccECSTaskDefinition_arrays
=== CONT  TestAccECSTaskDefinition_changeVolumesForcesNewResource
=== CONT  TestAccECSTaskDefinition_constraint
=== CONT  TestAccECSTaskDefinition_withRuntimePlatform
=== CONT  TestAccECSTaskDefinition_withTransitEncryptionEFSVolume
--- PASS: TestAccECSTaskDefinition_Fargate_ephemeralStorage (48.15s)
=== CONT  TestAccECSTaskDefinition_withPidMode
--- PASS: TestAccECSTaskDefinition_arrays (48.34s)
=== CONT  TestAccECSTaskDefinition_withIPCMode
--- PASS: TestAccECSTaskDefinition_inferenceAccelerator (48.40s)
=== CONT  TestAccECSTaskDefinition_withDockerVolume
--- PASS: TestAccECSTaskDefinition_withRuntimePlatform (49.05s)
=== CONT  TestAccECSTaskDefinition_withDockerVolumeMinimal
--- PASS: TestAccECSTaskDefinition_withTaskScopedDockerVolume (49.75s)
=== CONT  TestAccECSTaskDefinition_withEFSVolumeMinimal
--- PASS: TestAccECSTaskDefinition_withTaskRoleARN (52.59s)
=== CONT  TestAccECSTaskDefinition_withScratchVolume
--- PASS: TestAccECSTaskDefinition_constraint (52.77s)
=== CONT  TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch
--- PASS: TestAccECSTaskDefinition_withNetworkMode (52.81s)
=== CONT  TestAccECSTaskDefinition_Fargate_withRuntimePlatform
--- PASS: TestAccECSTaskDefinition_executionRole (53.41s)
--- PASS: TestAccECSTaskDefinition_withEFSVolume (61.99s)
--- PASS: TestAccECSTaskDefinition_proxy (63.02s)
--- PASS: TestAccECSTaskDefinition_withTransitEncryptionEFSVolume (63.48s)
--- PASS: TestAccECSTaskDefinition_fargate (65.93s)
--- PASS: TestAccECSTaskDefinition_withEFSAccessPoint (68.91s)
--- PASS: TestAccECSTaskDefinition_disappears (69.03s)
--- PASS: TestAccECSTaskDefinition_basic (74.58s)
--- PASS: TestAccECSTaskDefinition_changeVolumesForcesNewResource (75.26s)
--- PASS: TestAccECSTaskDefinition_withDockerVolume (32.76s)
--- PASS: TestAccECSTaskDefinition_withDockerVolumeMinimal (32.29s)
--- PASS: TestAccECSTaskDefinition_withScratchVolume (29.29s)
--- PASS: TestAccECSTaskDefinition_withIPCMode (33.55s)
--- PASS: TestAccECSTaskDefinition_withPidMode (33.93s)
--- PASS: TestAccECSTaskDefinition_Fargate_withRuntimePlatform (29.90s)
--- PASS: TestAccECSTaskDefinition_Fargate_withRuntimePlatformWithoutArch (30.10s)
--- PASS: TestAccECSTaskDefinition_withEFSVolumeMinimal (38.15s)
--- PASS: TestAccECSTaskDefinition_tags (90.74s)
--- PASS: TestAccECSTaskDefinition_withECSService (133.30s)
--- PASS: TestAccECSTaskDefinition_withFSxWinFileSystem (3138.02s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/ecs	3141.532s
```
